app/models/result.rb

```diff
@@ -204,7 +204,7 @@ def create_marks
   def mark_hash
     Hash[
       marks.map do |mark|
-        ["#{mark.markable_type}-#{mark.markable_id}",
+        ["#{mark.criterion.type}-#{mark.criterion_id}",
```

Comment:
This is another place where you should be able to remove the `criterion.type` part (check usages!).